### PR TITLE
Framebuffer

### PR DESCRIPTION
**Pull request by ashusharmasigdev**
All functions are surely working.

### CODE REVIEW

**Review by ashusharmasigdev — COMMENTED** [review]
Checked!

**Review by kotovalexarian — CHANGES_REQUESTED** [review]
First of all, thank you very much for the contribution!

Don't worry about the big number of requested changes. Most of them are more related to the code style, and I can fix them myself.

My only really important concern about this code is that it only supports one mode of framebuffer - graphical, and doesn't support text mode. While text mode is the feature of framebuffer I need mostly. Text mode is very useful for kernel debugging. And drivers in this library (console, framebuffer, maybe other in future like USB) are used mostly for debugging helpers, not as real drivers.

Can you at least add a support for mode change? We get framebuffer mode from a bootloader. Multiboot 2 and Stivale 2 boot protocols provide information about the mode. See `include/kernaux/multiboot2.h` for Multiboot 2 framebuffer info tag:

https://github.com/tailix/libkernaux/blob/5296e9cce3419f122a8ed72abdb4e0eb79dac2e9/include/kernaux/multiboot2.h#L136-L149

**Comment by ashusharmasigdev** [comment]
**Namaste** , I done the code as per your indications, this is my first contribution, so....
One more thing, I don't used much EGA or VGA , so some bugs can alive, maybe you need to fix those or i will in feature.
I uses C++ and Classes, so those are bit confused me in C.

**Comment by kotovalexarian** [comment]
@ashusharmasigdev, please give me access to your repo so I can edit the code before merge.

**Comment by ashusharmasigdev** [comment]
Access Granted.

**Comment by kotovalexarian** [comment]
I've fixed coding style, but text mode is still required. I'll add it later or you can do it youself.

File: src/framebuffer.c [hunk 1]
```diff
@@ -1,8 +1,92 @@
...
 #include "config.h"
 #endif
 
 #include <kernaux/framebuffer.h>
 
-__attribute__((unused))
-static const int foobar = 0;
+framebuffer_t framebuffer(uint32_t width, uint32_t height, uint32_t pitch, uint32_t depth, uintptr_t buf_addr)
+{
+    return (framebuffer_t){
```
[the remark this answers is not in the document]
Comment (kotovalexarian) [reply src/framebuffer.c]: Maybe more checks are needed.

**Comment by kotovalexarian** [comment]
@ashusharmasigdev Stivale 2 parser feature should be added in separate pull request.

**Comment by ashusharmasigdev** [comment]
@kotovalexarian What about adding a Basic VGA Font header file, to put it on framebuffer?

**Comment by kotovalexarian** [comment]
> @kotovalexarian What about adding a Basic VGA Font header file, to put it on framebuffer?

Let's finish the framebuffer itself first.

**Comment by kotovalexarian** [comment]
Hello again. I decided to remove drivers from **libkernaux** (#127) and create a new repository for them (have not created yet). Therefore I'm closing this PR. I'll write a driver for framebuffer later. This still isn't a priority because I use serial console for I/O in my projects, it's enough for development and testing.

**Comment by ashusharmasigdev** [comment]
Hello to you also, i currently can't continue my OS developments i am busy on some private web projects, i may continue later on,, this was the reason i can't contribute the libkernaux development. Do whatever your planing is. I may continue it later on.
Thanks.

**Comment by kotovalexarian** [comment]
Drivers will be here: https://github.com/tailix/drivers

**Pull request closed without merging**